### PR TITLE
issue-58--Calendar-Module

### LINKED ISSUE #58: Idea: Calendar Module
In case of having a "Community Calendar" we could integrate this to the Api, to access Data from that Calendar in other projects as well.
Before realizing that we should definitly specify some more Details - Feel free to Add Ideas in the Comments!

### LINKED ISSUE #75: Use Helmet
We should use helmet to improve security for our api.
Its like a default-package when building an api with express.
[Helmet](https://docs.nestjs.com/security/helmet)

### PR DESCRIPTION
**Pull request by Cahllagerfeld**
fixes #58 
fixes #75 

Calendar Module for saving community Events

- [x] CRUD Operations are ready.
- [x] e2e-Tests still missing

### CODE REVIEW

**Comment by Cahllagerfeld** [comment]
- [x] CRUD Operations are ready.
- [ ] e2e-Tests still missing

**Comment by Cahllagerfeld** [comment]
image
I implemented the response for the `getAll`-Endpoint like shown above:
- The Events are already sorted in future and ongoing, so the client doesnt need to implement this logic.
- When using a database we should define the timeranges before making the request and not filtering like here. Otherwise we have to read **ALL** Entries for events that have ever been made and this may be a huge payload.

**Comment by eddiejaoude** [comment]
Looking great 💥

**Comment by Cahllagerfeld** [comment]
Thanks :D I'll finish the e2e-Tests tomorrow.
I think when we got this module, theres just the database-thing open until we can make a release with the minimal set of requirements.
Are there any other feature-requests yet for future releases?

**Comment by eddiejaoude** [comment]
> Are there any other feature-requests yet for future releases?

For the calendar or generally?

**Comment by Cahllagerfeld** [comment]
I thought about general.
But I realized that this database stuff might be a bigger thing, than I initially thought, So I probably should rather focus on our open issues thanon future ones 😅

**Comment by Cahllagerfeld** [comment]
Dont merge this yet pls 😄 
I found a little bug regarding the validation pipe.
The tests are passing for now but I need to add something :D

**Comment by Cahllagerfeld** [comment]
I added the validationpipe for every dto.
Im currently implementing this validationpipe for the e2e testing which means I have to change response body for validation-errors.

My question now is:
When we use the `put`-Endpoint. Do we want to transfer the same object as when creating the entity, or do we just want to transfer the properties that have changed?

**Comment by eddiejaoude** [comment]
> When we use the put-Endpoint. Do we want to transfer the same object as when creating the entity, or do we just want to transfer the properties that have changed?

Great question! For a `put` it should be the whole object, any missing fields should be deleted from the DB. We should use `patch` to only update the sent fields and leave the other as they are

**Comment by Cahllagerfeld** [comment]
Okay, then fortunately I dont need to change that much 😄 
I'll open an issue for Patch then

**Comment by Cahllagerfeld** [comment]
Added the Validation Pipe.
I think this PR is ready for review then :)

File: src/main.ts
Comment (Cahllagerfeld) [inline src/main.ts:12]: Helmet is added here

Comment (eddiejaoude) [reply src/main.ts:12]: Awesome 👍

File: test/features/calendar.feature
Comment (eddiejaoude) [inline test/features/calendar.feature:46]: Do we also need a test to update a non-existing event?

File: test/features/calendar.feature
Comment (eddiejaoude) [inline test/features/calendar.feature:166]: I think we need to create another step def similar to this one that would make it easier to read - I can do it in another PR after this one is merged.

```ts
@then(/the response in item "([^"]*)" should contains:/)
public dataResponseItemTable(item: number, table: { rawTable: [] }) {
    const data = this.context.tableToObject(table);
    expect(JSON.parse(this.context.response.text)[item]).to.to.eql(data);
}
```
https://github.com/EddieHubCommunity/api/blob/main/test/step-definitions/responses.ts#L31

The new step def would be something like this

```ts
@then(/the response property "([^"]*)" in field "([^"]*)" should contains:/)
public dataResponseItemTable(property: string, item: number, table: { rawTable: [] }) {
    const data = this.context.tableToObject(table);
    expect(JSON.parse(this.context.response.text)[property][item]).to.eql(data);
}
```

The tests would be something like

```gherkin
And  the response property "future" in item "0" should contains:
            | id   | 123                |
            | name | "Livestream XY"    |
And  the response property "ongoing" in item "0" should contains:
            | id   | 123                |
            | name | "Livestream YZ"    |
```

Note: I think `contains` should be singular 🤔 - this can be changed later

**Comment by eddiejaoude** [comment]
Looks great 👍 just [1 comment](https://github.com/EddieHubCommunity/api/pull/76#pullrequestreview-644109991) about 1 more test

My other comment I can do in another PR after this one, it was just for some information

**Comment by Cahllagerfeld** [comment]
> Looks great +1 just [1 comment](https://github.com/EddieHubCommunity/api/pull/76#pullrequestreview-644109991) about 1 more test
> 
> My other comment I can do in another PR after this one, it was just for some information

Okay I'll add this test :smile:

**Review by eddiejaoude — APPROVED** [review]
Looks great, awesome work, thank you 👍

**Comment by Cahllagerfeld** [comment]
You're welcome :D I enjoyed it as always 😄

**Pull request merged**
